google/cloud/aiplatform/_pipeline_based_service/pipeline_based_service.py

```diff
@@ -0,0 +1,328 @@
+
+    client_class = utils.PipelineJobClientWithOverride
+    _resource_noun = "pipelineJob"
+    _delete_method = "delete_pipeline_job"
+    _getter_method = "get_pipeline_job"
+    _list_method = "list_pipeline_jobs"
+    _parse_resource_name_method = "parse_pipeline_job_path"
+    _format_resource_name_method = "pipeline_job_path"
+
+    _valid_done_states = _PIPELINE_COMPLETE_STATES
+
+    @property
+    @classmethod
+    @abc.abstractmethod
+    def _template_ref(self) -> Dict[str, str]:
+        """A dictionary of the pipeline template URLs for this servicewhere the key is
+        an identifier for that template and the value is the url of that pipeline template.
+
+        For example: {"tabular_classification": "gs://path/to/tabular/pipeline/template.json"}
+
+        """
+        pass
+
+    @property
+    @classmethod
+    @abc.abstractmethod
+    def _creation_log_message(self) -> str:
+        """A log message to use when the Pipeline-based Service is created, since this class
+        supresses logs from PipelineJob creation to avoid duplication.
+
+        For example: 'Created PipelineJob for your Model Evaluation.'
+
+        """
+        pass
+
+    # TODO: Consider updating this to return a dict or list in the future to support multiple outputs
+    @property
+    @abc.abstractmethod
+    def _metadata_output_artifact(self) -> Optional[str]:
+        """The ML Metadata output artifact resource URI from the completed pipeline run."""
+        pass
+
+    @property
+    def backing_pipeline_job(self) -> "pipeline_jobs.PipelineJob":
+        """The PipelineJob associated with the resource."""
+        return pipeline_jobs.PipelineJob.get(resource_name=self.resource_name)
+
+    @property
+    def pipeline_console_uri(self) -> Optional[str]:
+        """The console URI of the PipelineJob created by the service."""
+        if self.backing_pipeline_job:
+            return self.backing_pipeline_job._dashboard_uri()
+
+    @property
+    def state(self) -> Optional[gca_pipeline_state.PipelineState]:
+        """The state of the Pipeline run associated with the service."""
+        if self.backing_pipeline_job:
+            return self.backing_pipeline_job.state
+        return None
+
+    # TODO: expose _template_ref in error message when artifact registry support is added
+    def _validate_pipeline_template_matches_service(
+        self, pipeline_job: "pipeline_jobs.PipelineJob"
+    ):
+        """Utility function to validate that the passed in pipeline matches
+        the template of the Pipeline Based Service.
+
+        Args:
+            pipeline_job (aiplatform.PipelineJob):
+                Required. The PipelineJob to validate with this Pipeline Based Service.
+
+        Raises:
+            ValueError: if the provided pipeline ID doesn't match the pipeline service.
+        """
+
+        pipeline_match = False
+
+        for pipeline_template in self._template_ref.values():
```

Comment:
If i read correctly, this compares the serialized `component` fields.
You should probably compare pipeline_job.to_dict()["pipelineSpec"] instead since it's the template. The template also contains the root component which is important to decide the topology of the components.
PipelineSpec is the IR